### ISSUE TITLE
Critical: Missing Data Preprocessing Components and Model Training Pipeline Failures

### ISSUE THREAD

**Issue by aaron-seq** [aaron-seq/BPpredictiontool#8]
## Problem Description

The data preprocessing pipeline and model training infrastructure have critical issues that prevent proper data processing, model training, and prediction generation. Multiple missing components and configuration errors will cause runtime failures.

### 1. Missing Preprocessor Initialization

**Problem in main.py:**
```python
global model_loaded
global preprocessor_fitted

model_loaded = False
preprocessor_fitted = False  # Never set to True anywhere
```

**Issues:**
- `setup_preprocessor()` function is called but never sets `preprocessor_fitted = True`
- Predictions will fail because preprocessor state is never validated
- No error handling for preprocessor failures

### 2. Incomplete Data Preprocessing Module

**Missing from backend/data_processing/:**
- Only `advanced_preprocessor.py` exists
- No basic preprocessing utilities
- No data validation functions
- No signal quality assessment implementation
- Missing feature extraction modules

**Expected but Missing:**
```python
# These imports in main.py will fail
from data_processing import basic_preprocessor
from data_processing import feature_extractor
from data_processing import signal_validator
```

### 3. Model Path Configuration Problems

**Config.py Issues:**
```python
class DataConfiguration:
    models_directory: Path = Path('models')  # Relative path - problematic
```

**Problems:**
- Relative paths fail in Docker containers
- No validation that model files exist
- Model loading attempts to access non-existent paths:
  - `data/models_directory/advanced_bp_model.keras`
  - `data/models_directory/blood_pressure_model.keras`

**Actual Issue:**
- Model files are in `backend/models/` directory
- Config points to `data/models/` directory
- Path mismatch causes model loading to fail

### 4. Missing Training Data and Scripts

**No Training Infrastructure:**
- No training data directory or sample data
- No model training scripts
- No data augmentation utilities
- No hyperparameter tuning configuration
- Cannot retrain or update models

**Missing Critical Files:**
- `backend/train_model.py`
- `backend/train_advanced_model.py`
- `backend/data/` directory with training samples
- Model evaluation and validation scripts

### 5. Preprocessor Class Implementation Issues

**advanced_preprocessor.py Problems:**
```python
class AdvancedPreprocessor:
    def __init__(self, sequence_length=1000):
        self.sequence_length = sequence_length
        self.scaler = None  # Never initialized
        self.is_fitted = False  # Never set to True
```

**Critical Issues:**
- Scaler is None but used in transform operations
- No fit() method called before transform()
- Signal quality assessment not fully implemented
- Missing wavelet transform functionality referenced
- No normalization strategy defined

### 6. Data Validation Missing

**No Input Validation:**
```python
@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()
    ppg_signal = np.array(data.get('ppg_signal'))  # No validation
    ecg_signal = np.array(data.get('ecg_signal'))  # No validation
```

**Missing Validations:**
- Signal length validation
- Data type checking
- NaN/Inf value detection
- Signal quality thresholds
- Amplitude range validation
- Sampling rate verification

### 1. Incomplete Signal Processing

**Missing Implementations:**
- Bandpass filtering for noise reduction
- Artifact removal algorithms
- Peak detection validation
- R-peak detection for ECG
- PPG waveform quality metrics

### 2. No Model Version Management

**Issues:**
- No model versioning system
- Cannot rollback to previous models
- No A/B testing capability
- No model performance tracking
- Missing model metadata

### 3. Configuration Hardcoding

**Hardcoded Values:**
```python
sequence_length = 1000  # Should be from config
sampling_rate = 125     # Hardcoded, no flexibility
num_features = 2        # PPG + ECG, but not configurable
```

### 1. Fix Preprocessor Initialization

**Update backend/main.py:**
```python
def setup_preprocessor():
    """Setup the data preprocessor."""
    global preprocessor_fitted
    
    try:
        if config.data.data_directory.exists():
            # Initialize and fit preprocessor
            from data_processing.advanced_preprocessor import AdvancedPreprocessor
            
            preprocessor = AdvancedPreprocessor(
                sequence_length=config.model.sequence_length
            )
            
            # Check if we have training data to fit
            training_data_path = config.data.data_directory / 'training_samples.npz'
            if training_data_path.exists():
                training_data = np.load(training_data_path)
                preprocessor.fit(training_data['signals'])
                preprocessor_fitted = True
                logger.info("Preprocessor fitted successfully")
            else:
                logger.warning("No training data found for preprocessor fitting")
                preprocessor_fitted = False
                
        return True
    except Exception as e:
        logger.error(f"Error setting up preprocessor: {str(e)}")
        return False
```

### 2. Create Missing Data Processing Components

**Create backend/data_processing/__init__.py:**
```python
"""Data processing module for signal preprocessing and feature extraction."""

from .advanced_preprocessor import AdvancedPreprocessor
from .signal_validator import SignalValidator
from .feature_extractor import FeatureExtractor

__all__ = ['AdvancedPreprocessor', 'SignalValidator', 'FeatureExtractor']
```

**Create backend/data_processing/signal_validator.py:**
```python
"""Signal quality validation and assessment."""
import numpy as np
from typing import Tuple, Dict

class SignalValidator:
    def __init__(self, min_quality_threshold: float = 0.7):
        self.min_quality_threshold = min_quality_threshold
    
    def validate_signal(self, signal: np.ndarray) -> Tuple[bool, Dict]:
        """Validate signal quality and return metrics."""
        quality_metrics = {
            'has_nan': np.isnan(signal).any(),
            'has_inf': np.isinf(signal).any(),
            'length': len(signal),
            'std': np.std(signal),
            'mean': np.mean(signal)
        }
        
        is_valid = (
            not quality_metrics['has_nan'] and
            not quality_metrics['has_inf'] and
            quality_metrics['length'] >= 1000 and
            quality_metrics['std'] > 0
        )
        
        return is_valid, quality_metrics
```

### 3. Fix Model Path Configuration

**Update backend/config.py:**
```python
from pathlib import Path
import os

@dataclass
class DataConfiguration:
    # Use absolute paths for production
    base_dir: Path = Path(__file__).parent
    data_directory: Path = base_dir / 'data'
    models_directory: Path = base_dir / 'models'  # FIXED: Now points to backend/models
    logs_directory: Path = base_dir / 'logs'
    min_signal_quality_threshold: float = 0.7
    normalization_method: str = 'minmax'
```

### 4. Add Input Validation

**Update backend/main.py prediction route:**
```python
from data_processing.signal_validator import SignalValidator

validator = SignalValidator(config.data.min_signal_quality_threshold)

@app.route('/predict', methods=['POST'])
def predict():
    try:
        data = request.get_json()
        
        # Validate input data exists
        if not data or 'ppg_signal' not in data or 'ecg_signal' not in data:
            return jsonify({
                "error": "Missing required fields: ppg_signal and ecg_signal"
            }), 400
        
        # Convert to numpy arrays
        ppg_signal = np.array(data.get('ppg_signal'))
        ecg_signal = np.array(data.get('ecg_signal'))
        
        # Validate signals
        ppg_valid, ppg_metrics = validator.validate_signal(ppg_signal)
        ecg_valid, ecg_metrics = validator.validate_signal(ecg_signal)
        
        if not ppg_valid:
            return jsonify({
                "error": "Invalid PPG signal",
                "details": ppg_metrics
            }), 400
            
        if not ecg_valid:
            return jsonify({
                "error": "Invalid ECG signal", 
                "details": ecg_metrics
            }), 400
        
        # Continue with prediction...
        
    except ValueError as e:
        return jsonify({"error": f"Invalid data format: {str(e)}"}), 400
```

### 5. Create Training Script

**Create backend/train_model.py:**
```python
"""Model training script for blood pressure prediction."""
import numpy as np
from models.advanced_lstm_model import AttentionBloodPressureModel
from data_processing.advanced_preprocessor import AdvancedPreprocessor
from config import ApplicationConfig

def train_model():
    config = ApplicationConfig()
    
    # Load training data
    data_path = config.data.data_directory / 'training_data.npz'
    if not data_path.exists():
        raise FileNotFoundError(f"Training data not found at {data_path}")
    
    data = np.load(data_path)
    X_train = data['signals']
    y_train = data['labels']
    
    # Initialize model
    model = AttentionBloodPressureModel(config)
    bp_model = model.build_model(input_shape=(config.model.sequence_length, 2))
    
    # Train model
    history = bp_model.fit(
        X_train, y_train,
        batch_size=config.model.batch_size,
        epochs=config.model.epochs,
        validation_split=config.model.validation_split
    )
    
    # Save model
    model_path = config.data.models_directory / 'advanced_bp_model.keras'
    bp_model.save(model_path)
    print(f"Model saved to {model_path}")

if __name__ == '__main__':
    train_model()
```

## Impact Assessment

- **Severity**: Critical
- **Component**: Data Processing Pipeline, Model Training
- **Functionality**: Predictions will fail, cannot retrain models
- **Data Quality**: No validation leads to incorrect predictions
- **Maintainability**: Cannot update or improve models

## Files That Must Be Created

**New Files Required:**
- `backend/data_processing/__init__.py`
- `backend/data_processing/signal_validator.py`
- `backend/data_processing/feature_extractor.py`
- `backend/train_model.py`
- `backend/train_advanced_model.py`
- `backend/data/training_data.npz` (sample data)
- `backend/evaluate_model.py`

**Files to Update:**
- `backend/config.py` (fix model paths)
- `backend/main.py` (add validation, fix preprocessor)
- `backend/data_processing/advanced_preprocessor.py` (complete implementation)

## Testing Required

- [ ] Verify preprocessor can be initialized and fitted
- [ ] Test signal validation with valid and invalid inputs
- [ ] Confirm model paths resolve correctly
- [ ] Test prediction endpoint with various input formats
- [ ] Verify error handling for malformed data
- [ ] Test model training script
- [ ] Validate preprocessor transformation produces expected output
- [ ] Check signal quality assessment accuracy

This issue prevents the core ML functionality from working properly and blocks the ability to improve or retrain models.
